### PR TITLE
Fix calibrate powder bs

### PR DESCRIPTION
**Pull request by obejarano**

### CODE REVIEW

**Comment by obejarano** [comment]
still missing is the DL MISO "fix" needed to remove CFO. script is stable but performance is poor.

File: MATLAB/iris_py.py [hunk 1]
```diff
@@ -193,7 +193,7 @@ def sdr_settxfreq(self, tx_freq):
 
         def sdr_settxgain(self, tx_gain):
                 for chan in [0, 1]:
-                    self.sdr.setGain(SOAPY_SDR_TX, chan, min(tx_gain, 81.0))
+                    self.sdr.setGain(SOAPY_SDR_TX, chan, tx_gain) #min(tx_gain, 81.0))
```
Comment (asedlmayr) [inline MATLAB/iris_py.py:196]: remove the comment

File: MATLAB/mimo_driver.m [hunk 2]
```diff
@@ -46,208 +48,99 @@
...
                     ue_id_str(1, i) = {convertStringsToChars(obj.ue_serial_ids(i))};
                 end
                 ue_id_list = py.list(ue_id_str);
 
+                % Reference Node (overkill... it will always be at most one)
+                ref_id_str = cell(1, n_ref_sdrs);
+                for i = 1:n_ref_sdrs
+                    ref_id_str(1, i) = {convertStringsToChars(obj.ref_serial_ids(i))};
+                end
+                ref_id_list = py.list(ref_id_str);
```
Comment (asedlmayr) [inline MATLAB/mimo_driver.m:77]: This is screaming to make a function..............

File: MATLAB/mimo_driver.py [hunk 3]
```diff
@@ -261,17 +269,19 @@ def txrx_downlink(self, tx_data_mat_re, tx_data_mat_im, num_frames, nsamps_pad=1
...
-        [bs.config_sdr_tdd(tdd_sched=str(bs_sched_b) if i == 0 else str(bs_sched), nsamps=n_samps, prefix_len=nsamps_pad, dualpilot=dual_pilot) for i, bs in enumerate(self.bs_obj)]
+        [bs.config_sdr_tdd(tdd_sched=str(bs_sched_b) if i == 0 else str(bs_sched), nsamps=n_samps, prefix_len=nsamps_pad) for i, bs in enumerate(self.bs_obj)]
         [ue.config_sdr_tdd(is_bs=False, tdd_sched=str(ue_sched[i]), nsamps=n_samps, prefix_len=nsamps_pad) for i, ue in enumerate(self.ue_obj)]
 
+        #for i, bs in enumerate(self.bs_obj):
+        #    bs.burn_data_complex(tx_data_mat[i, :] if n_bs_antenna > 1 else tx_data_mat)
+
         for i, bs in enumerate(self.bs_obj):
-            bs.burn_data_complex(tx_data_mat[i, :] if n_bs_antenna > 1 else tx_data_mat)
+            #bs.burn_data_complex(tx_data_mat[i, :] if n_bs_antenna > 1 else tx_data_mat) # for matlab
```
Comment (asedlmayr) [inline MATLAB/mimo_driver.py:280]: remove commented code

File: MATLAB/mimo_driver.py [hunk 4]
```diff
@@ -392,16 +410,25 @@ def txrx_dl_sound(self, tx_data_mat_re, tx_data_mat_im, num_frames, nsamps_pad=1
...
 
+        #savemat("./testtest.mat", {'tx_data_mat': tx_data_mat});
         for i, bs in enumerate(self.bs_obj):
             if python_mode:
-                bs.burn_data_complex(tx_data_mat[0])  # For python
+                bs.burn_data_complex(tx_data_mat[0])  # For python (needs more testing)
             else:
-                bs.burn_data_complex(tx_data_mat)     # for matlab
-
+                #bs.burn_data_complex(tx_data_mat[i, :] if n_bs_antenna > 1 else tx_data_mat) # for matlab
```
Comment (asedlmayr) [inline MATLAB/mimo_driver.py:427]: Removed commented code

File: MATLAB/mimo_driver.py [hunk 5]
```diff
@@ -392,16 +410,25 @@ def txrx_dl_sound(self, tx_data_mat_re, tx_data_mat_im, num_frames, nsamps_pad=1
...
+            dual_pilot = True
+        else:
+            dual_pilot = False
+
         ### Write TDD Schedule ###
-        [bs.config_sdr_tdd(tdd_sched=str(bs_sched[i]), nsamps=n_samps, prefix_len=nsamps_pad) for i, bs in enumerate(self.bs_obj)]
+        [bs.config_sdr_tdd(tdd_sched=str(bs_sched[i]), nsamps=n_samps, prefix_len=nsamps_pad, dualpilot=dual_pilot) for i, bs in enumerate(self.bs_obj)]
         [ue.config_sdr_tdd(is_bs=False, tdd_sched=str(ue_sched[i]), nsamps=n_samps, prefix_len=nsamps_pad) for i, ue in enumerate(self.ue_obj)]
 
+        #savemat("./testtest.mat", {'tx_data_mat': tx_data_mat});
```
Comment (asedlmayr) [inline MATLAB/mimo_driver.py:422]: Removed commented code

File: MATLAB/mimo_driver.py [hunk 6]
```diff
@@ -482,27 +509,38 @@ def txrx_refnode(self, tx_data_mat_re, tx_data_mat_im, num_frames, bs_sched, ue_
...
-            rx_data[frame, :, :, :] = np.reshape(np.array(rx_data_frame[0][0]), (self.n_bs_antenna, numRxSyms, n_samps))
+            #rx_data[frame, i, :, :] = np.reshape(np.array(rx_data_frame[0][i]), (self.n_bs_antenna, numRxSyms, n_samps))
+            for i, bs in enumerate(self.bs_obj):
+                if self.n_bs_chan > 1:
+                    rx_data[frame, i*2, :, :] = np.reshape(np.array(rx_data_frame[i][0]), (numRxSyms, n_samps))
+                    rx_data[frame, i*2+1, :, :] = np.reshape(np.array(rx_data_frame[i][1]), (numRxSyms, n_samps))
+                else:
+                    rx_data[frame, i, :, :] = np.reshape(np.array(rx_data_frame[i][0]), (numRxSyms, n_samps))
  
+        #axs[1].plot(np.real(np.squeeze(rx_data[0,0,0,:])))
```
Comment (asedlmayr) [inline MATLAB/mimo_driver.py:541]: Removed commented code

File: MATLAB/rl_ofdm_siso.m [hunk 7]
```diff
@@ -185,7 +189,7 @@
         'tx_gain_ue', TX_GN_UE, ...
         'rxgain', RX_GN, ...
         'sample_rate', SMPL_RT, ...
-	'trig_offset', TX_ADVANCE);
+	    'trig_offset', TX_ADVANCE);
```
Comment (asedlmayr) [inline MATLAB/rl_ofdm_siso.m:192]: whitespace off

File: MATLAB/rl_ofdm_ul_mimo.m [hunk 8]
```diff
@@ -30,10 +30,12 @@
 close all;
 
 
-[version, executable, isloaded] = pyversion;
-if ~isloaded
-    pyversion /usr/bin/python
-    py.print() % weird bug where py isn't loaded in an external script
+%[version, executable, isloaded] = pyversion;
+pe = pyenv;
+%disp(pe);
```
Comment (asedlmayr) [inline MATLAB/rl_ofdm_ul_mimo.m:35]: Removed commented code

File: PYTHON/DEMOS/SISO_RX.py [hunk 9]
```diff
@@ -75,7 +75,8 @@
 from fft_power import *
 from macros import *
 from init_fncs import *
-from MyFuncAnimation import *
+from scipy.io import savemat
+#from MyFuncAnimation import *
```
Comment (asedlmayr) [inline PYTHON/DEMOS/SISO_RX.py:79]: Removed commented code

File: PYTHON/DEMOS/SISO_RX.py [hunk 10]
```diff
@@ -268,10 +269,9 @@ def rxsamples_app(srl, freq, gain, num_samps, recorder, agc_en, wait_trigger):
     setUpDigitalRssiMode(sdr)
 
     # There's a bug in the FuncAnimation function, we replaced it with a fixed version
-    # anim = animation.FuncAnimation(fig, animate, init_func=init, fargs=(num_samps, recorder, agc_en, wait_trigger),
-    # frames=100, interval=100, blit=True)
-    anim = MyFuncAnimation(fig, animate, init_func=init, fargs=(num_samps, recorder, agc_en, wait_trigger, info),
-                                    frames=100, interval=100, blit=True)
+    anim = animation.FuncAnimation(fig, animate, init_func=init, fargs=(num_samps, recorder, agc_en, wait_trigger, info),frames=100, interval=100, blit=True)
+    #anim = MyFuncAnimation(fig, animate, init_func=init, fargs=(num_samps, recorder, agc_en, wait_trigger, info),
```
Comment (asedlmayr) [inline PYTHON/DEMOS/SISO_RX.py:273]: Remove commented code

File: PYTHON/DEMOS/SISO_RX.py [hunk 11]
```diff
@@ -322,6 +322,9 @@ def animate(i, num_samps, recorder, agc_en, wait_trigger, info):
 
     # Store received samples in binary file (second method of storage)
     write_to_file('./data_out/rxsamps', sampsRx)
+    #mdic = {"rxdata": sampsRx[0]}
+    #savemat("data_rmode2.mat", mdic)
```
Comment (asedlmayr) [inline PYTHON/DEMOS/SISO_RX.py:326]: Remove commented code

**Review by asedlmayr — APPROVED** [review]
Removed commented code

File: MATLAB/mimo_driver.m (hunk 2 above)
Comment (obejarano) [reply MATLAB/mimo_driver.m:77]: will do it on the next commit

File: PYTHON/DEMOS/SISO_RX.py (hunk 11 above)
Comment (obejarano) [reply PYTHON/DEMOS/SISO_RX.py:326]: useful for testing things, I'll try to make it prettier on the next commit

**Pull request merged**